bentoml/_internal/runner/local.py

```diff
@@ -40,6 +41,11 @@ def run(self, *args: t.Any, **kwargs: t.Any) -> t.Any:
                 batch_result,
                 batch_axis=self._runner.batch_options.output_batch_axis,
             )
+            if isinstance(single_results, GeneratorType):
```

Comment:
It would be great to move this check to the `DefaultContainer#batch_to_singles`, I think it is nice if the default Data Container can handle generator type.  For other Runner data container types(Numpy, Pandas, etc) we probably won't need this check here.

Thanks @eob for the PR! We haven't got around to document the design of Runner and DataContainer yet, basically Runner class implements a `run_batch` method, allow runner instance to invoke `run` or `run_batch` call. These calls will be dynamically batched, in order to better utilize hardware concurrency.  In order to do this, BentoML needs to know how to group a batch of `run` inputs into a batch data type for the `_run_batch` method, and the DataContainer class is meant for definition that process for each data type that Runner supports.